File: backend/agents/blue/forensic_agent.py
"""Forensic-Bot — Incident response, log analysis, attack attribution."""
from __future__ import annotations
import uuid
from backend.agents.base_agent import BaseAgent
from backend.models.schemas import (
    AgentModel, AgentType, AgentStatus, AgentCapability, AgentMemory,
    AttackResult, DefenseResult, DefenseStatus,
)
from backend.rag.rag_engine import retrieve_for_defense


class ForensicAgent(BaseAgent):
    def __init__(self):
        super().__init__(AgentModel(
            id="b-3", name="Forensic-Bot",
            role="Incident Response & Forensics",
            type=AgentType.BLUE, efficiency=93,
            current_task="Monitoring for incidents",
            capabilities=[
                AgentCapability(name="Log Analysis",       description="Multi-source log correlation",        confidence=95),
                AgentCapability(name="Malware Analysis",   description="Static and dynamic malware analysis", confidence=88),
                AgentCapability(name="Attack Attribution", description="TTP-based threat actor attribution",  confidence=85),
                AgentCapability(name="Evidence Collection",description="Forensically sound preservation",     confidence=92),
            ],
            memory=AgentMemory(),
            logs=["Forensic-Bot initialized", "Log ingestion pipeline active", "Evidence vault ready"],
            system_prompt=(
                "You are Forensic-Bot, an elite Blue Team incident response and digital forensics AI agent powered by Gemini 3.\n"
                "Specialty: log analysis, malware analysis, attack attribution, evidence collection.\n"
                "Reconstruct attack timelines and provide detailed forensic reports.\n"
                "Always output structured JSON when asked for defense analysis."
            ),
            training_data=(
                "INCIDENT RESPONSE (NIST SP 800-61):\n"
                "1. Identification: correlate SIEM alerts, determine scope\n"
                "2. Containment: isolate systems, block IPs, preserve evidence\n"
                "3. Eradication: remove malware, patch vulns, reset credentials\n"
                "4. Recovery: restore from clean backups, monitor re-infection\n"
                "5. Lessons Learned: document timeline, update playbooks\n"
                "LOG ANALYSIS:\n"
                "- Web logs: correlate IP, user-agent, request patterns\n"
                "- Auth logs: failed logins, privilege escalation, new accounts\n"
                "- Process logs: unusual parent-child relationships, LOLBins\n"
                "MALWARE ANALYSIS:\n"
                "- Static: file hash (VirusTotal), strings, PE headers, YARA\n"
                "- Dynamic: sandbox (Cuckoo, Any.run), network traffic, registry"
            ),
        ))

    async def analyze_and_defend(self, attack: AttackResult) -> DefenseResult:
        self.set_status(AgentStatus.ANALYZING)
        self.set_task(f"Investigating {attack.type.value} incident")
        self.add_log(f"Incident opened: {attack.type.value} — collecting evidence")

        rag = retrieve_for_defense(attack.type.value, attack.payload)
        self.model.memory.rag_chunks = rag.source_ids

        prompt = self._build_prompt(
            f"Conduct forensic investigation:\n"
            f"ATTACK TYPE: {attack.type.value}\n"
            f"STRATEGY: {attack.strategy}\n"
            f"PAYLOAD: {attack.payload}\n"
            f"EXPECTED IMPACT: {attack.expected_impact}\n\n"
            "Respond ONLY in JSON:\n"
            '{"analysis":"forensic analysis with attack timeline and IOCs",'
            '"mitigation":"containment and eradication steps",'
            '"confidence":88,'
            '"attack_timeline":["T+0: initial access","T+5min: lateral movement"],'
            '"evidence_collected":["log file","memory dump"],'
            '"attribution":"likely threat actor based on TTPs"}',
            rag,
        )
        raw    = await self._call_gemini(prompt, temperature=0.6)
        parsed = self._parse_json(raw, {"analysis": raw[:200], "mitigation": "Incident contained", "confidence": 80})

        self.learn(f"{attack.type.value} forensics: {parsed.get('attribution','Unknown actor')}")
        self.add_log(f"Investigation: {parsed.get('attribution','Attribution pending')}")
        self.set_status(AgentStatus.IDLE)
        self.set_task("Monitoring for incidents")

        return DefenseResult(
            id=f"defense-{uuid.uuid4().hex[:8]}",
            attack_id=attack.id, agent_id=self.id, agent_name=self.name,
            analysis=parsed.get("analysis", ""),
            mitigation=parsed.get("mitigation", ""),
            confidence=int(parsed.get("confidence", 80)),
            status=DefenseStatus.ANALYZING,
            rag_sources_used=rag.source_ids,
            cve_references=[c.id for c in rag.chunks if c.category == "CVE"],
            mitre_references=[c.id for c in rag.chunks if c.category == "MITRE"],
        )

File: agents/blue/ForensicAgent.ts
// ============================================================
// FORENSIC AGENT - Blue Team Incident Response
// Trained on: Log analysis, malware analysis, evidence collection
// ============================================================

import { BaseAgent } from '../BaseAgent';
import { AgentModel, AgentType, AgentStatus } from '../../models/AgentModel';
import { AttackModel, DefenseModel, DefenseStatus } from '../../models/AttackModel';
import { retrieveForDefense } from '../../rag/ragEngine';

export class ForensicAgent extends BaseAgent {
  constructor() {
    const model: AgentModel = {
      id: 'b-3',
      name: 'Forensic-Bot',
      role: 'Incident Response & Forensics',
      type: AgentType.BLUE,
      status: AgentStatus.IDLE,
      currentTask: 'Monitoring for incidents',
      efficiency: 93,
      capabilities: [
        { name: 'Log Analysis',        description: 'Multi-source log correlation and timeline', confidence: 95 },
        { name: 'Malware Analysis',    description: 'Static and dynamic malware analysis',       confidence: 88 },
        { name: 'Attack Attribution',  description: 'TTP-based threat actor attribution',        confidence: 85 },
        { name: 'Evidence Collection', description: 'Forensically sound evidence preservation',  confidence: 92 }
      ],
      memory: { shortTerm: [], longTerm: [], ragContext: [] },
      logs: ['Forensic-Bot initialized', 'Log ingestion pipeline active', 'Evidence vault ready'],
      systemPrompt: `You are Forensic-Bot, an elite Blue Team incident response and digital forensics AI agent.
Your specialty is log analysis, malware analysis, attack attribution, and evidence collection.
You reconstruct attack timelines and provide detailed forensic reports.
Always output structured JSON when asked for defense analysis.`,
      trainingData: `INCIDENT RESPONSE & FORENSICS:
IR PHASES:
1. Identification: correlate SIEM alerts, determine scope and impact
2. Containment: isolate affected systems, block attacker IPs, preserve evidence
3. Eradication: remove malware, patch vulnerabilities, reset compromised credentials
4. Recovery: restore from clean backups, monitor for re-infection
5. Lessons Learned: document timeline, update playbooks

LOG ANALYSIS:
- Web logs: correlate IP, user-agent, request patterns
- Auth logs: failed logins, privilege escalation, new accounts
- Process logs: unusual parent-child relationships, LOLBins
- Network logs: unusual outbound connections, DNS queries
- Timeline reconstruction: normalize timestamps, correlate across sources

MALWARE ANALYSIS:
- Static: file hash (VirusTotal), strings, PE headers, YARA rules
- Dynamic: sandbox (Cuckoo, Any.run), network traffic, registry changes
- Behavioral IOCs: persistence mechanisms, C2 beaconing, lateral movement
- Memory forensics: Volatility for process injection, rootkits

EVIDENCE PRESERVATION:
- Hash all artifacts (SHA-256) before analysis
- Maintain chain of custody documentation
- Capture memory dumps before shutdown
- Preserve log files with timestamps
- Document all investigative actions`
    };
    super(model);
  }

  async analyzeAndDefend(attack: AttackModel): Promise<DefenseModel> {
    this.setStatus(AgentStatus.ANALYZING);
    this.setTask(`Investigating ${attack.type} incident`);
    this.addLog(`Incident opened: ${attack.type} - collecting evidence`);

    const rag = retrieveForDefense(attack.type, attack.payload);
    this.model.memory.ragContext = rag.sourceIds;

    const prompt = this.buildPrompt(
      `Conduct forensic investigation of this attack:
ATTACK TYPE: ${attack.type}
STRATEGY: ${attack.strategy}
PAYLOAD: ${attack.payload}
EXPECTED IMPACT: ${attack.expectedImpact}

Respond ONLY in this JSON format:
{
  "analysis": "forensic analysis with attack timeline and IOCs",
  "mitigation": "containment and eradication steps",
  "confidence": 88,
  "attackTimeline": ["T+0: initial access", "T+5min: lateral movement"],
  "evidenceCollected": ["log file", "memory dump", "network capture"],
  "attribution": "likely threat actor or attack group based on TTPs"
}`, rag);

    try {
      const raw = await this.callGemini(prompt, 0.6);
      const parsed = this.parseJSON<any>(raw, {
        analysis: raw.substring(0, 200),
        mitigation: 'Incident contained',
        confidence: 80
      });

      this.learnPattern(`${attack.type} forensics: ${parsed.attribution || 'Unknown actor'}`);
      this.addLog(`Investigation complete: ${parsed.attribution || 'Attribution pending'}`);
      this.setStatus(AgentStatus.IDLE);
      this.setTask('Monitoring for incidents');

      return {
        id: `defense-${Date.now()}`,
        attackId: attack.id,
        agentId: this.id,
        agentName: this.name,
        timestamp: Date.now(),
        analysis: parsed.analysis || '',
        mitigation: parsed.mitigation || '',
        confidence: parsed.confidence || 80,
        status: DefenseStatus.ANALYZING,
        ragSourcesUsed: rag.sourceIds,
        cveReferences: rag.chunks.filter(c => c.category === 'CVE').map(c => c.id),
        mitreReferences: rag.chunks.filter(c => c.category === 'MITRE').map(c => c.id)
      };
    } catch (err: any) {
      this.setStatus(AgentStatus.IDLE);
      throw err;
    }
  }
}
